core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala

```diff
@@ -0,0 +1,343 @@
+   * Shutdown this [[RpcEnv]] asynchronously. If need to make sure [[RpcEnv]] exits successfully,
+   * call [[awaitTermination()]] straight after [[shutdown()]].
+   */
+  def shutdown(): Unit
+
+  /**
+   * Wait until [[RpcEnv]] exits.
+   *
+   * TODO do we need a timeout parameter?
+   */
+  def awaitTermination(): Unit
+
+  /**
+   * Create a URI used to create a [[RpcEndpointRef]]
+   */
+  def uriOf(systemName: String, address: RpcAddress, endpointName: String): String
+}
+
+private[spark] case class RpcEnvConfig(
+    conf: SparkConf,
+    name: String,
+    host: String,
+    port: Int,
+    securityManager: SecurityManager)
+
+/**
+ * A RpcEnv implementation must have a companion object with an
+ * `apply(config: RpcEnvConfig): RpcEnv` method so that it can be created via Reflection.
+ *
+ * {{{
+ * object MyCustomRpcEnv {
+ *   def apply(config: RpcEnvConfig): RpcEnv = {
+ *     ...
+ *   }
+ * }
+ * }}}
+ */
+private[spark] object RpcEnv {
+
+  private def getRpcEnvCompanion(conf: SparkConf): AnyRef = {
+    // Add more RpcEnv implementations here
+    val rpcEnvNames = Map("akka" -> "org.apache.spark.rpc.akka.AkkaRpcEnv")
+    val rpcEnvName = conf.get("spark.rpc", "akka")
+    val rpcEnvClassName = rpcEnvNames.getOrElse(rpcEnvName.toLowerCase, rpcEnvName)
+    val companion = Class.forName(
+      rpcEnvClassName + "$", true, Utils.getContextOrSparkClassLoader).getField("MODULE$").get(null)
+    companion
+  }
+
+  def create(
+       name: String,
+       host: String,
+       port: Int,
+       conf: SparkConf,
+       securityManager: SecurityManager): RpcEnv = {
+    // Using Reflection to create the RpcEnv to avoid to depend on Akka directly
+    val config = RpcEnvConfig(conf, name, host, port, securityManager)
+    val companion = getRpcEnvCompanion(conf)
+    companion.getClass.getMethod("apply", classOf[RpcEnvConfig]).
+      invoke(companion, config).asInstanceOf[RpcEnv]
+  }
+
+}
+
+/**
+ * An end point for the RPC that defines what functions to trigger given a message.
+ *
+ * RpcEndpoint will be guaranteed that `onStart`, `receive` and `onStop` will
+ * be called in sequence.
+ *
+ * The lift-cycle will be:
+ *
+ * constructor onStart receive* onStop
+ *
+ * If any error is thrown from one of RpcEndpoint methods except `onError`, [[RpcEndpoint.onError]]
+ * will be invoked with the cause. If onError throws an error, [[RpcEnv]] will ignore it.
+ */
+private[spark] trait RpcEndpoint {
+
+  /**
+   * The [[RpcEnv]] that this [[RpcEndpoint]] is registered to.
+   */
+  val rpcEnv: RpcEnv
+
+  /**
+   * Provide the implicit sender. `self` will become valid when `onStart` is called.
+   *
+   * Note: Because before `onStart`, [[RpcEndpoint]] has not yet been registered and there is not
+   * valid [[RpcEndpointRef]] for it. So don't call `self` before `onStart` is called. In the other
+   * words, don't call [[RpcEndpointRef.send]] in the constructor of [[RpcEndpoint]].
+   */
+  implicit final def self: RpcEndpointRef = {
```

Comment:
is there a case in which we would ever want a different sender? if not, i'd say just remove this and always use the self sendor